bytecomp/lambda.ml

```diff
@@ -788,10 +793,13 @@ let rec map f lam =
 
 (* To let-bind expressions to variables *)
 
-let bind str var exp body =
+let bind_with_value_kind str (var, kind) exp body =
   match exp with
     Lvar var' when Ident.same var var' -> body
-  | _ -> Llet(str, Pgenval, var, exp, body)
+  | _ -> Llet(str, kind, var, exp, body)
+
+let bind str var exp body =
+  bind_with_value_kind str (var, Pgenval) exp body
```

Comment:
I've looked at the use points for this function. They are all in `Matching`. Most of them do have kind information. I think it would probably be better to update the uses rather than the definition.